src/erc721.rs

```diff
@@ -0,0 +1,367 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+use sp_runtime::traits::Hash as HashT;
+use support::{
+    decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchError,
+    dispatch::DispatchResult, ensure, traits::Randomness,
+};
+use system::ensure_signed;
+
+// Encoding library
+use codec::Encode;
+
+pub trait Trait: system::Trait + balances::Trait + contracts::Trait {
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+    // Something that provides randomness generation in the runtime.
+    type Randomness: Randomness<Self::Hash>;
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        // No owner set for token
+        TokenNoOwner,
+
+        // Got an overflow after adding
+        Overflow,
+
+        // Got an underflow after subing
+        Underflow,
+    }
+}
+
+decl_event!(
+    pub enum Event<T> 
+        where
+        <T as system::Trait>::AccountId,
+        <T as system::Trait>::Hash {
+
+        // Token transfer event
+        Transfer(Option<AccountId>, Option<AccountId>, Hash),
+
+        // One token approved to an account
+        Approval(AccountId, AccountId, Hash),
+
+        // All tokens owned are approved to other account
+        ApprovalForAll(AccountId, AccountId, bool),
+    }
+);
+
+decl_storage! {
+    trait Store for Module<T: Trait> as ERC721 {
+        // Token count owned by an account
+        pub OwnedTokensCount get(balance_of): map T::AccountId => u64;
+
+        // Token owner's account id
+        pub TokenOwner get(owner_of): map T::Hash => Option<T::AccountId>;
+
+        // Token delegate's account id
+        pub TokenApprovals get(get_approved): map T::Hash => Option<T::AccountId>;
+
+        // Account deledate's account id
+        pub OperatorApprovals get(is_approved_for_all): map (T::AccountId, T::AccountId) => bool;
+
+        // Total count of minted token
+        pub TotalSupply get(total_supply): u64;
+
+        // Map of token index to token id
+        pub AllTokens get(token_by_index): map u64 => T::Hash;
```

Comment:
1. I think u64 is enough for index of token. In my opioion, the main reason for uint256 in Ethereum is the EVM storage model, its byte length is 256 bit. 
2. Enumerable is important for Dapp. I have developed projects based on Ethereum before. The UI sometimes need list all minted token, or list all tokens owned by an account.

Reply:
@juniuszhou you are right with the EVM storage with 256 bits but I guess for the hash you are using some kind of uint256 otherwise it won't work.

What would be the downside if we use 256 bits everywhere?

Reply:
Can we make somehow a modular implementation? Base is ERC721 and if required we can use ERC721Enumerable?

Reply:
1. For Hash type in ERC721, it can be set when build the runtime. It should be H256 for sure. For total supply and index, I think u64 is enough. It seems no benefits if set length as 256 bits.
2. I will redesign the ERC721 as separate deliverable pallet. It is just a mod behind the NFT now. To my experience, enumerable is important feature in real world. For example, some system use ERC721 as digital evidence, then most of tokens owned by several major accounts. I am thinking there maybe problem if use vector to store one account's owned tokens.

Reply:
I'm fine with u64, but please add a note to the README of ERC721 so that users are aware of that change.